azkaban-common/src/main/java/azkaban/metric/inmemoryemitter/InMemoryMetricEmitter.java

```diff
@@ -0,0 +1,233 @@
+  private static final String INMEMORY_METRIC_REPORTER_WINDOW = "azkaban.metric.inmemory.interval";
+  private static final String INMEMORY_METRIC_NUM_INSTANCES = "azkaban.metric.inmemory.maxinstances";
+  private static final String INMEMORY_METRIC_STANDARDDEVIATION_FACTOR =
+      "azkaban.metric.inmemory.standardDeviationFactor";
+
+  private double _standardDeviationFactor;
+  /**
+   * Interval (in millisecond) from today for which we should maintain the in memory snapshots
+   */
+  private long _timeWindow;
+  /**
+   * Maximum number of snapshots that should be displayed on /stats servlet
+   */
+  private long _numInstances;
+
+  /**
+   * @param azkProps Azkaban Properties
+   */
+  public InMemoryMetricEmitter(Props azkProps) {
+    _historyListMapping = new HashMap<String, LinkedList<InMemoryHistoryNode>>();
+    _timeWindow = azkProps.getLong(INMEMORY_METRIC_REPORTER_WINDOW, 60 * 60 * 24 * 7 * 1000);
+    _numInstances = azkProps.getLong(INMEMORY_METRIC_NUM_INSTANCES, 50);
+    _standardDeviationFactor = azkProps.getDouble(INMEMORY_METRIC_STANDARDDEVIATION_FACTOR, 2);
+  }
+
+  /**
+   * Update reporting interval
+   * @param val interval in milli seconds
+   */
+  public synchronized void setReportingInterval(long val) {
+    _timeWindow = val;
+  }
+
+  /**
+   * Set number of /stats servlet display points
+   * @param num
+   */
+  public void setReportingInstances(long num) {
+    _numInstances = num;
+  }
+
+  /**
+   * Ingest metric in snapshot data structure while maintaining interval
+   * {@inheritDoc}
+   * @see azkaban.metric.IMetricEmitter#reportMetric(azkaban.metric.IMetric)
+   */
+  @Override
+  public void reportMetric(final IMetric<?> metric) throws Exception {
+    String metricName = metric.getName();
+    if (!_historyListMapping.containsKey(metricName)) {
+      _logger.info("First time capturing metric: " + metricName);
+      _historyListMapping.put(metricName, new LinkedList<InMemoryHistoryNode>());
+    }
+    synchronized (_historyListMapping.get(metricName)) {
+      _logger.debug("Ingesting metric: " + metricName);
+      _historyListMapping.get(metricName).add(new InMemoryHistoryNode(metric.getValue()));
+      cleanUsingTime(metricName, _historyListMapping.get(metricName).peekLast().getTimestamp());
+    }
+  }
+
+  /**
+   * Get snapshots for a given metric at a given time
+   * @param metricName name of the metric
+   * @param from Start date
+   * @param to end date
+   * @param useStats get statistically significant points only
+   * @return List of snapshots
+   */
+  public List<InMemoryHistoryNode> getDrawMetric(final String metricName, final Date from, final Date to,
+      final Boolean useStats) throws ClassCastException {
+    LinkedList<InMemoryHistoryNode> selectedLists = new LinkedList<InMemoryHistoryNode>();
+    if (_historyListMapping.containsKey(metricName)) {
+
+      _logger.debug("selecting snapshots within time frame");
+      synchronized (_historyListMapping.get(metricName)) {
```

Comment:
I used synchronized to avoid simultaneous read and write on metric history for a particular metric.  getDrawMetric is called from a an executor service so it should not be blocking other azkaban functions ?